### PR TITLE
Adding JavaSMT support

### PR DESCRIPTION
**Pull request by leventeBajczi**
This PR adds support for the JavaSMT framework, thus enabling a long list of solvers to be natively used by Theta.

Furthermore, there are some patches to the CI checks as to not fail due to memory limits, and cvc5 libraries are also included in the lib folder from now on. Unfortunately, no cvc5 support is present in JavaSMT for Windows/Mac for now.

There were also of course some bugs I uncovered, especially regarding floating point operations. I tried my best in fixing them.

I also created further testFixtures for expressions, and I include them in non-JavaSMT unit tests as well.

### CODE REVIEW

File: subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/fptype/FpType.java [hunk 1]
```diff
@@ -68,7 +68,7 @@ public int getSignificand() {
 
     @Override
     public EqExpr<FpType> Eq(Expr<FpType> leftOp, Expr<FpType> rightOp) {
-        return FpEqExpr.of(leftOp, rightOp);
+        return FpEqExpr.of(leftOp, rightOp); // TODO: shouldn't this be assign?
```
[the remark this answers is not in the document]
Comment (leventeBajczi) [reply subprojects/common/core/src/main/java/hu/bme/mit/theta/core/type/fptype/FpType.java:71]:
```suggestion
        return FpAssignExpr(leftOp, rightOp);
```

File: subprojects/solver/solver-javasmt/src/main/java/hu/bme/mit/theta/solver/javasmt/JavaSMTExprTransformer.java [hunk 2]
```diff
@@ -0,0 +1,1266 @@
...
+                expr.getRightOp());
+        return integerFormulaManager.modulo(leftOpTerm, rightOpTerm);
+    }
+
+    private Formula transformIntRem(final IntRemExpr expr) {
+        final IntegerFormula leftOpTerm = (IntegerFormula) toTerm(
+                expr.getLeftOp());
+        final IntegerFormula rightOpTerm = (IntegerFormula) toTerm(
+                expr.getRightOp());
+        return integerFormulaManager.modulo(leftOpTerm, rightOpTerm); // TODO: this is mod, not rem. Is this OK?
```
[the remark this answers is not in the document]
Comment (leventeBajczi) [reply subprojects/solver/solver-javasmt/src/main/java/hu/bme/mit/theta/solver/javasmt/JavaSMTExprTransformer.java:724]: Although we should really leave REM alone, and not use it anywhere else either.

File: subprojects/solver/solver-javasmt/src/main/java/hu/bme/mit/theta/solver/javasmt/JavaSMTExprTransformer.java [hunk 3]
```diff
@@ -0,0 +1,1266 @@
...
+        final BitvectorFormula rightOpTerm = (BitvectorFormula) toTerm(expr.getRightOp());
+
+        return bitvectorFormulaManager.shiftRight(leftOpTerm, rightOpTerm, true); // TODO: cross-check signedness
+    }
+
+    private Formula transformBvLogicShiftRight(final BvLogicShiftRightExpr expr) {
+        final BitvectorFormula leftOpTerm = (BitvectorFormula) toTerm(expr.getLeftOp());
+        final BitvectorFormula rightOpTerm = (BitvectorFormula) toTerm(expr.getRightOp());
+
+        return bitvectorFormulaManager.shiftRight(leftOpTerm, rightOpTerm, false); // TODO: cross-check signedness
```
Comment (leventeBajczi) [inline subprojects/solver/solver-javasmt/src/main/java/hu/bme/mit/theta/solver/javasmt/JavaSMTExprTransformer.java:945]: Is signedness == arithmetic/logical?
I might need some help. 
@as3810t ?

File: subprojects/solver/solver-javasmt/src/main/java/hu/bme/mit/theta/solver/javasmt/JavaSMTExprTransformer.java [hunk 4]
```diff
@@ -0,0 +1,1266 @@
...
+            literal[i + 1 + expr.getType().getExponent()] = significand[i];
+        }
+        return context.getFormulaManager().getUFManager().declareAndCallUF("to_fp",
+                FormulaType.getFloatingPointType(expr.getType().getExponent(), expr.getType().getSignificand() - 1),
+                bitvectorFormulaManager.makeBitvector(
+                        size,
+                        BvUtils.neutralBvLitExprToBigInteger(BvLitExpr.of(literal))),
+                integerFormulaManager.makeNumber(expr.getType().getExponent()),
+                integerFormulaManager.makeNumber(expr.getType().getSignificand() - 1)
+        ); // TODO: this is an uninterpreted function right now. Will not work properly.
```
Comment (leventeBajczi) [inline subprojects/solver/solver-javasmt/src/main/java/hu/bme/mit/theta/solver/javasmt/JavaSMTExprTransformer.java:1043]:
```suggestion
        );
```

File: subprojects/solver/solver-javasmt/src/main/java/hu/bme/mit/theta/solver/javasmt/JavaSMTExprTransformer.java [hunk 5]
```diff
@@ -0,0 +1,1266 @@
...
+
+    private Formula transformArrayLit(final ArrayLitExpr<?, ?> expr) {
+        throw new JavaSMTSolverException("Array literals not yet supported: " + expr);
+
+    }
+
+    private Formula transformArrayInit(final ArrayInitExpr<?, ?> expr) {
+        throw new JavaSMTSolverException("Array literals not yet supported: " + expr);
+    }
+
```
Comment (leventeBajczi) [inline subprojects/solver/solver-javasmt/src/main/java/hu/bme/mit/theta/solver/javasmt/JavaSMTExprTransformer.java:1252]: https://github.com/sosy-lab/java-smt/issues/360 will hopefully bring about a fix for these.

File: subprojects/solver/solver-javasmt/src/main/java/hu/bme/mit/theta/solver/javasmt/JavaSMTTermTransformer.java [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+
+    private Expr<?> transformBvLit(final Formula term, BigInteger value) {
+        final var bvNum = (BitvectorFormula) term;
+        BitvectorType formulaType = (BitvectorType) context.getFormulaManager().getFormulaType(bvNum);
+
+        return BvUtils.bigIntegerToNeutralBvLitExpr(value, formulaType.getSize());
+    }
+
+    private Expr<?> transformFpLit(final Formula term, BigDecimal value) {
+        final var fpNum = (FloatingPointFormula) term;
```
Comment (leventeBajczi) [inline subprojects/solver/solver-javasmt/src/main/java/hu/bme/mit/theta/solver/javasmt/JavaSMTTermTransformer.java:337]: If https://github.com/sosy-lab/java-smt/issues/360 fixes this, we should implement it correctly.

File: subprojects/solver/solver-z3-legacy/src/main/java/hu/bme/mit/theta/solver/z3legacy/Z3ExprTransformer.java [hunk 7]
```diff
@@ -1160,7 +1160,7 @@ private com.microsoft.z3legacy.Expr transformFpFromBv(final FpFromBvExpr expr) {
         final FPSort fpSort = context.mkFPSort(expr.getFpType().getExponent(),
                 expr.getFpType().getSignificand());
         return context.mkFPToFP(transformFpRoundingMode(expr.getRoundingMode()), val, fpSort,
-                expr.isSigned());
+                expr.isSigned()); // TODO: is this OK? FP2FP when BV2FP is used?
```
Comment (leventeBajczi) [inline subprojects/solver/solver-z3-legacy/src/main/java/hu/bme/mit/theta/solver/z3legacy/Z3ExprTransformer.java:1163]: @as3810t , do you know anything about this?

File: subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3ExprTransformer.java [hunk 8]
```diff
@@ -1160,7 +1160,7 @@ private com.microsoft.z3.Expr transformFpFromBv(final FpFromBvExpr expr) {
         final FPSort fpSort = context.mkFPSort(expr.getFpType().getExponent(),
                 expr.getFpType().getSignificand());
         return context.mkFPToFP(transformFpRoundingMode(expr.getRoundingMode()), val, fpSort,
-                expr.isSigned());
+                expr.isSigned());  // TODO: is this OK? FP2FP when BV2FP is used?
```
Comment (leventeBajczi) [inline subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3ExprTransformer.java:1163]: Same as the other z3 transformer: @as3810t , do you know anything about this?

File: subprojects/solver/solver-z3/src/test/java/hu/bme/mit/theta/solver/z3/Z3UserPropagatorTest.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
+/*
```
Comment (leventeBajczi) [inline subprojects/solver/solver-z3/src/test/java/hu/bme/mit/theta/solver/z3/Z3UserPropagatorTest.java:1]: Remove this file.

**Comment by leventeBajczi** [comment]
I addressed all TODOs I left for myself, so I consider this PR ready. @as3810t, when you have the time, please take a look!

**Review by as3810t — APPROVED** [review]
Ithink it is a great addition to the existing solvers.

Questions:
- Right now, CVC5 and JavaSMT dependencies are included directly as a `jar` into the project. Can we use Gradle's dependency management instead?

Suggestions:
- It would be great to add description on how to use the new solvers. I get that the names start with `JavaSMT:...`, but it would be useful to document the different strings in the README that can follow the colon.

**Pull request merged**